File: packages/capacitor-plugin/android/src/main/java/com/otakit/updater/EventOutbox.java
package com.otakit.updater;

import android.util.AtomicFile;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONArray;
import org.json.JSONObject;

/** A bounded, atomic snapshot. Transport never owns or changes event identity. */
final class EventOutbox {

  static final int LIMIT = 256;
  static final long TTL = 7L * 24 * 60 * 60 * 1000;

  static final class Entry {

    final String id, url, appId, body;
    final long createdAt, nextAt;
    final int attempts;

    Entry(String url, String appId, String body, long createdAt, int attempts, long nextAt)
      throws Exception {
      this.id = new JSONObject(body).getString("eventId");
      this.url = url;
      this.appId = appId;
      this.body = body;
      this.createdAt = createdAt;
      this.attempts = attempts;
      this.nextAt = nextAt;
    }

    JSONObject json() throws Exception {
      return new JSONObject()
        .put("url", url)
        .put("appId", appId)
        .put("body", body)
        .put("createdAt", createdAt)
        .put("attempts", attempts)
        .put("nextAt", nextAt);
    }
  }

  private final AtomicFile file;
  private List<Entry> entries = new ArrayList<>();

  EventOutbox(File directory) throws Exception {
    if (!directory.isDirectory() && !directory.mkdirs()) throw new IOException(
      "Cannot create event outbox"
    );
    file = new AtomicFile(new File(directory, "events.json"));
    if (!file.getBaseFile().exists() && !new File(file.getBaseFile() + ".bak").exists()) return;
    byte[] bytes;
    try (java.io.InputStream input = file.openRead()) {
      if (file.getBaseFile().length() > 4 * 1024 * 1024) throw new IOException(
        "Event outbox too large"
      );
      java.io.ByteArrayOutputStream buffer = new java.io.ByteArrayOutputStream();
      byte[] chunk = new byte[8192];
      int count;
      while ((count = input.read(chunk)) != -1) {
        if (buffer.size() + count > 4 * 1024 * 1024) throw new IOException(
          "Event outbox too large"
        );
        buffer.write(chunk, 0, count);
      }
      bytes = buffer.toByteArray();
    }
    try {
      JSONArray stored = new JSONArray(new String(bytes, StandardCharsets.UTF_8));
      for (int i = 0; i < stored.length(); i++) {
        JSONObject item = stored.getJSONObject(i);
        entries.add(
          new Entry(
            item.getString("url"),
            item.getString("appId"),
            item.getString("body"),
            item.getLong("createdAt"),
            item.getInt("attempts"),
            item.getLong("nextAt")
          )
        );
      }
      if (entries.size() > LIMIT) throw new IllegalStateException("Invalid outbox length");
    } catch (Exception corrupted) {
      android.util.Log.w("OtaKit", "Discarding unreadable event outbox");
      save(new ArrayList<>());
    }
  }

  synchronized void enqueue(String url, String appId, String body, long now) throws Exception {
    if (body.getBytes(StandardCharsets.UTF_8).length > 8192) throw new IOException(
      "Device event too large"
    );
    List<Entry> next = live(now);
    Entry entry = new Entry(url, appId, body, now, 0, now);
    for (Entry existing : next) if (existing.id.equals(entry.id)) return;
    while (next.size() >= LIMIT) next.remove(0);
    next.add(entry);
    save(next);
  }

  synchronized Entry ready(long now) throws Exception {
    List<Entry> next = live(now);
    if (next.size() != entries.size()) save(next);
    for (Entry entry : entries) if (entry.nextAt <= now) return entry;
    return null;
  }

  synchronized Long waitMilliseconds(long now) {
    Long delay = null;
    for (Entry entry : entries) {
      long due = Math.min(entry.nextAt, entry.createdAt + TTL);
      long candidate = Math.max(0, due - now);
      delay = delay == null ? candidate : Math.min(delay, candidate);
    }
    return delay;
  }

  synchronized void complete(String id, int status, String retryAfter, long now, double random)
    throws Exception {
    List<Entry> next = live(now);
    for (int i = 0; i < next.size(); i++) {
      Entry entry = next.get(i);
      if (!entry.id.equals(id)) continue;
      if (
        (status >= 200 && status < 300) ||
        (status >= 300 && status < 500 && status != 408 && status != 425 && status != 429)
      ) {
        next.remove(i);
      } else {
        int attempts = Math.min(entry.attempts + 1, 20);
        long backoff = (long) (Math.min(3_600_000, 5000L << Math.min(attempts - 1, 10)) *
          (0.5 + random));
        Long serverDelay = DownloadRetry.retryAfterMilliseconds(retryAfter, now);
        long delay = serverDelay == null ? backoff : Math.max(backoff, serverDelay);
        long due = delay > Long.MAX_VALUE - now ? Long.MAX_VALUE : now + delay;
        next.set(i, new Entry(entry.url, entry.appId, entry.body, entry.createdAt, attempts, due));
      }
      break;
    }
    save(next);
  }

  private List<Entry> live(long now) {
    List<Entry> result = new ArrayList<>();
    for (Entry entry : entries) if (now < entry.createdAt + TTL) result.add(entry);
    return result;
  }

  private void save(List<Entry> next) throws Exception {
    JSONArray data = new JSONArray();
    for (Entry entry : next) data.put(entry.json());
    FileOutputStream output = null;
    byte[] bytes = data.toString().getBytes(StandardCharsets.UTF_8);
    try {
      output = file.startWrite();
      output.write(bytes);
      output.getFD().sync();
      file.finishWrite(output);
      output = null;
      if (!java.util.Arrays.equals(bytes, file.readFully())) throw new IOException(
        "Event outbox commit failed"
      );
      entries = next;
    } catch (Exception error) {
      if (output != null) file.failWrite(output);
      throw error;
    }
  }
}

File: packages/capacitor-plugin/ios/Sources/UpdaterPlugin/EventOutbox.swift
import Foundation

/// Atomic, bounded persistence; retries preserve the exact original event body.
final class EventOutbox {
  static let limit = 256
  static let ttl: TimeInterval = 7 * 24 * 60 * 60
  struct Entry: Codable {
    let id: String
    let url: URL
    let appId: String
    let body: Data
    let createdAt: TimeInterval
    var attempts: Int
    var nextAt: TimeInterval
  }
  private let file: URL
  private let lock = NSLock()
  private var entries: [Entry] = []

  init(directory: URL) throws {
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    var directory = directory
    var values = URLResourceValues()
    values.isExcludedFromBackup = true
    try directory.setResourceValues(values)
    file = directory.appendingPathComponent("events.json")
    if FileManager.default.fileExists(atPath: file.path) {
      let size = try file.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? 0
      guard size <= 4 * 1024 * 1024 else { throw CocoaError(.fileReadTooLarge) }
      let data = try Data(contentsOf: file)
      do {
        entries = try JSONDecoder().decode([Entry].self, from: data)
        guard entries.count <= Self.limit else { throw CocoaError(.fileReadCorruptFile) }
      } catch {
        print("[OtaKit] Discarding unreadable event outbox")
        try save([])
      }
    }
  }

  func enqueue(url: URL, appId: String, body: Data, now: TimeInterval) throws {
    lock.lock(); defer { lock.unlock() }
    guard body.count <= 8192,
          let payload = try JSONSerialization.jsonObject(with: body) as? [String: Any],
          let id = payload["eventId"] as? String else { throw CocoaError(.fileWriteInvalidFileName) }
    var next = live(now)
    guard !next.contains(where: { $0.id == id }) else { return }
    while next.count >= Self.limit { next.removeFirst() }
    next.append(Entry(id: id, url: url, appId: appId, body: body, createdAt: now, attempts: 0, nextAt: now))
    try save(next)
  }

  func ready(now: TimeInterval) throws -> Entry? {
    lock.lock(); defer { lock.unlock() }
    let next = live(now)
    if next.count != entries.count { try save(next) }
    return entries.first { $0.nextAt <= now }
  }

  func wait(now: TimeInterval) -> TimeInterval? {
    lock.lock(); defer { lock.unlock() }
    return entries.map { max(0, min($0.nextAt, $0.createdAt + Self.ttl) - now) }.min()
  }

  func complete(id: String, status: Int, retryAfter: String?, now: TimeInterval, random: Double) throws {
    lock.lock(); defer { lock.unlock() }
    var next = live(now)
    if let index = next.firstIndex(where: { $0.id == id }) {
      if (200..<300).contains(status) || ((300..<500).contains(status) && ![408, 425, 429].contains(status)) {
        next.remove(at: index)
      } else {
        next[index].attempts = min(next[index].attempts + 1, 20)
        let backoff = min(3600, 5 * pow(2, Double(min(next[index].attempts - 1, 10)))) * (0.5 + random)
        let serverDelay = DownloadRetry.retryAfter(retryAfter, now: Date(timeIntervalSince1970: now)) ?? 0
        // Expiration is the upper bound; never retry before a longer server delay.
        next[index].nextAt = min(now + max(backoff, serverDelay), next[index].createdAt + Self.ttl)
      }
    }
    try save(next)
  }

  private func live(_ now: TimeInterval) -> [Entry] {
    entries.filter { now < $0.createdAt + Self.ttl }
  }

  private func save(_ next: [Entry]) throws {
    try JSONEncoder().encode(next).write(to: file, options: .atomic)
    entries = next
  }
}

/// Owns one serial delivery loop independently of any Capacitor bridge lifetime.
final class EventDelivery: NSObject, URLSessionTaskDelegate {
  private let worker = DispatchQueue(label: "OtaKit.events")
  private let stateLock = NSLock()
  private var outbox: EventOutbox?
  private var scheduled: DispatchWorkItem?
  private var sending = false
  private let configuration: URLSessionConfiguration
  private lazy var session = URLSession(configuration: configuration, delegate: self, delegateQueue: nil)

  init(outbox: EventOutbox? = nil, configuration: URLSessionConfiguration = .ephemeral) {
    self.outbox = outbox
    self.configuration = configuration
    super.init()
  }

  func resume() {
    stateLock.lock()
    if outbox == nil {
      do {
        let support = try FileManager.default.url(for: .applicationSupportDirectory, in: .userDomainMask,
          appropriateFor: nil, create: true)
        outbox = try EventOutbox(directory: support.appendingPathComponent("OtaKitEvents", isDirectory: true))
      } catch { print("[OtaKit] Cannot open device event outbox") }
    }
    stateLock.unlock()
    worker.async { self.schedule() }
  }

  func enqueue(url: URL, appId: String, body: Data) {
    stateLock.lock()
    let outbox = self.outbox
    stateLock.unlock()
    do {
      guard let outbox else { throw CocoaError(.fileWriteUnknown) }
      try outbox.enqueue(url: url, appId: appId, body: body, now: Date().timeIntervalSince1970)
      worker.async { self.schedule() }
    } catch { print("[OtaKit] Cannot persist device event") }
  }

  private func schedule(minimumDelay: TimeInterval = 0) {
    guard !sending else { return }
    scheduled?.cancel()
    stateLock.lock(); let outbox = self.outbox; stateLock.unlock()
    guard let delay = outbox?.wait(now: Date().timeIntervalSince1970) else { scheduled = nil; return }
    let item = DispatchWorkItem { self.drain() }
    scheduled = item
    worker.asyncAfter(deadline: .now() + max(minimumDelay, delay), execute: item)
  }

  private func drain() {
    guard !sending else { return }
    scheduled?.cancel(); scheduled = nil
    stateLock.lock(); let outbox = self.outbox; stateLock.unlock()
    guard let outbox else { return }
    do {
      guard let entry = try outbox.ready(now: Date().timeIntervalSince1970) else { schedule(); return }
      sending = true
      var request = URLRequest(url: entry.url)
      request.httpMethod = "POST"
      request.setValue(entry.appId, forHTTPHeaderField: "X-App-Id")
      request.setValue("application/json", forHTTPHeaderField: "Content-Type")
      request.httpBody = entry.body
      request.timeoutInterval = 10
      session.dataTask(with: request) { _, response, error in
        self.worker.async {
          let http = response as? HTTPURLResponse
          var storageFailed = false
          do {
            try outbox.complete(id: entry.id, status: error == nil ? (http?.statusCode ?? 0) : 0,
              retryAfter: http?.value(forHTTPHeaderField: "Retry-After"),
              now: Date().timeIntervalSince1970, random: Double.random(in: 0..<1))
          } catch {
            storageFailed = true
            print("[OtaKit] Cannot persist device event delivery state")
          }
          self.sending = false
          self.schedule(minimumDelay: storageFailed ? 60 : 0)
        }
      }.resume()
    } catch {
      print("[OtaKit] Cannot read device event outbox")
      schedule(minimumDelay: 60)
    }
  }

  func urlSession(_ session: URLSession, task: URLSessionTask,
    willPerformHTTPRedirection response: HTTPURLResponse, newRequest request: URLRequest,
    completionHandler: @escaping (URLRequest?) -> Void) {
    completionHandler(nil)
  }
}
